### ISSUE TITLE
Blitting SVG Animations

### ISSUE THREAD

**Issue by jungerm2** [matplotlib/matplotlib#19694]
### Problem

I'm thrilled that SVG animations will be fully supported in the upcoming release. I've been doing some testing to see how it compares to other animation formats, and to my surprise, the resulting file size of an jshtml svg animation is huge! Here's a small test program that compares the filesize of animations:

```python
import sys
import numpy as np
from tqdm.auto import tqdm, trange
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from IPython.display import HTML

plt.rcParams['animation.embed_limit'] = 2**128

def anim_memfootprint(size, frame_format='png'):
    plt.rcParams['animation.frame_format'] = frame_format

    def update_line(num, data, line):
        line.set_data(data[..., :num])
        return line,
    fig1 = plt.figure()

    # Fixing random state for reproducibility
    np.random.seed(19680801)

    data = np.random.rand(2, size)
    l, = plt.plot([], [], 'r-')
    plt.xlim(0, 1)
    plt.ylim(0, 1)

    line_ani = animation.FuncAnimation(fig1, update_line, size, fargs=(data, l),
                                       interval=50, blit=True)
    footprint = sys.getsizeof(line_ani.to_jshtml())
    plt.close(fig1)
    
    return footprint

x = list(range(110, 5100, 100))
pngs =  [anim_memfootprint(i, 'png') for i in tqdm(x)]
svgs =  [anim_memfootprint(i, 'svg') for i in tqdm(x)]
jpegs =  [anim_memfootprint(i, 'jpeg') for i in tqdm(x)]

plt.plot(x, pngs, label='pngs')
plt.plot(x, svgs, label='svgs')
plt.plot(x, jpegs, label='jpegs')
plt.legend()
```  

If we plot this we get the following plot, with #points/frames on the X-axis and #bytes of the resulting animation on the Y-axis:

![image](https://user-images.githubusercontent.com/17375657/110992921-3d05d180-833c-11eb-9b34-4a3869cb3729.png)

The Y-axis is in 100s of MBs! Sure, there are up to 5000 frames and a max of 5000 points to plot, but still! 
If we consider that each point is an (x, y) pair, each 32bits then that's 5000x2x(32/8) = 40,000 bytes or about 0.04 MB of data to plot. Here, all the rest is static from frame to frame and therefore should be a constant overhead. 

### Proposed Solution

In these animations, each frame is stored separately and displayed as an image using an image tag with a data URL (where the image is passed in as a base64 encoded string). This makes sense if the format is something like PNG or JPEG as we can't easily only redraw part of it, but for SVG we totally can!

In fact, manipulating SVG via JS is very common. We could have a single SVG image and use JS to only show the parts that are relevant for each frame. In the case of the above, we could do this via the `dashoffset` SVG property as showcased [here](https://jakearchibald.com/2013/animated-line-drawing-svg/).

Blitting this way is not always possible as animations can do more than animate paths. But there are other SVG attributes that can be set dynamically that can be used to not have to redraw the whole frame every time. For instance, the SVG `visibility` attribute (see [here](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/visibility)) can be set as needed. This would enable SVG animations to only contain one "master" SVG document that is edited as needed. 

This is not necessarily complicated either, say there is very little overlap between frames because the animation shows a series of images. We'd have something like:
```svg
<svg width="200" height="200"  xmlns="http://www.w3.org/2000/svg">
  <!--  A ton of metadata goes here -->
  <!--  Setup axes, title, legends, etc here -->
  <image id="frame-1" href="..." height="200" width="200"/>
  <image id="frame-2" href="..." height="200" width="200"/>
  <image id="frame-3" href="..." height="200" width="200"/>
  <image id="frame-4" href="..." height="200" width="200"/>
  <!-- etc -->
</svg>
```

Then the JS which is responsible for currently updating that image tag I mentioned previously can simply set the visibility of the current frame to `visible` and everything else to `hidden`. This would enable re-use of the rest of the figure, and allow for a huge reduction in filesize. 

I'm sure there are other tricks we can play here, but the point is that with SVG animations we can blit and it'll make everything so much faster. I'm routinely generating animations that are 200MB which could be only a few MBs. 

And if we can't effectively perform this blitting, then we can simply fall back on the current behavior!

**Comment by tacaswell** [comment]
I am not sure I agree with using the term "blitting" for this (as I associate that term very strongly with raster graphics not vector graphics), but I definitely agree with the sentiment (that we should be using svg features to animate more intelligently)!

I think it would be a fair amount of work, but adding an option to `to_jshtml()` to  produce fully vector svgs which are internally animated would be great.

**Comment by jungerm2** [comment]
I agree, I couldn't think of a better word than blitting though! I've been trying to think about the best way to approach this problem and how to implement it, but I'm still familiarizing myself with the (enormous) codebase. 

Any ideas on how this could be achieved? Or rather at which level? 

Implementing this in the SVG backend would make the most sense, but in order to find the places that can be optimized away would require integration with the animation code. 

I'd love to see this feature (and maybe even contribute!), I just don't know how to start.

**Comment by jklymak** [comment]
Do we, or could we, add enough unique "id"s to SVG files so that objects could be tracked and duplicates found?  Then we could do this after making the svgs as a external step.  It seems that is an easier lift than trying to intertwine svg rendering with animation.  

For instance if I do 
```
import numpy as np
import matplotlib.pyplot as plt

fig, ax = plt.subplots( constrained_layout=True)
p = ax.plot(range(10))
print(p)
fig.savefig('boo.svg')
p[0].set_ydata(range(0, 20, 2))
fig.savefig('boo2.svg')
```

I get the same objects in each, though the urls seem to change:

```svg
<g id="line2d_11">
    <path clip-path="url(#pc3dc98c57e)" d="M 36.428059 311.24339 
L 81.017657 278.625067 
L 125.607255 246.006744 
L 170.196853 213.388422 
L 214.786451 180.770099 
L 259.376049 148.151776 
L 303.965647 115.533453 
L 348.555245 82.915131 
L 393.144843 50.296808 
L 437.734441 17.678485 
" style="fill:none;stroke:#1f77b4;stroke-linecap:square;stroke-width:1.5;"/>
   </g>
```

```svg 
<g id="line2d_11">
    <path clip-path="url(#p04d122ee68)" d="M 36.428059 311.24339 
L 81.017657 246.006744 
L 125.607255 180.770099 
L 170.196853 115.533453 
L 214.786451 50.296808 
L 249.848095 -1 
" style="fill:none;stroke:#1f77b4;stroke-linecap:square;stroke-width:1.5;"/>
   </g>
```
Writing something that just animated the changes doesn't seem too bad.  

(Yes, you'd still have all the svg files, but the merging step would be make a much compressed file)

**Comment by jungerm2** [comment]
Making this an external step would maybe make it easier to implement and yet, I can't help but wonder if there are already some built-in tools (in the animation or blitting modules) that we can take advantage of. Detecting duplicates and more generally overlaps might be easier (and much faster, since we wouldn't need to generate each frame) at this stage as compared to a post-processing stage.   

Doing this after the fact would involve reading each frame, and diffing them to find their difference. From there, the main SVG document would need to be built, ids changed, etc. And modifying the JS code will also be required. It seems easier to have a JS template (much like the one that already exists) for SVG animations, instead of generating code on the fly based on what needs animating. 

As for the URL thing you've pointed out, I'm not sure what's causing that, but as I recall the JS methods/ids/etc and HTML tags get assigned random unique ids to minimize conflicts if multiple jshtml animations are opened in the same document. So maybe that can explain the different URLs.

**Comment by jungerm2** [comment]
Looks like the unique URL is actually an ID that gets generated here (L671):
https://github.com/matplotlib/matplotlib/blob/0bc924d1cc8b17586e6f1d3bb76199dd6ba81cc2/lib/matplotlib/backends/backend_svg.py#L657-L677

**Comment by jungerm2** [comment]
Okay, so I was able to come up with a potential solution. It's a post-processing approach that diffs consecutive frames. Instead of storing all frames it only stores enough info to recompute them. It's definitely not what I originally had in mind, and could easily be improved further, but still, it shows significant filesize reduction. 

Here's a working PoC: https://github.com/jungerm2/HTMLDiffWriter  

Thoughts?

**Comment by tacaswell** [comment]
If you have an SVG reneder can you get it to "draw" just one Artist?  I know there is a lot of streaming logic in the svg backend, but not sure how deeply entwined the state is....

**Comment by jungerm2** [comment]
Ideally, an artist would include some state that records how that artist is changed over time, right? Then a collection of artists would more seamlessly be converted into an animation. From my understanding, this isn't the case and thus understanding the "framewise artist update" is quite complicated. 

If we had such an Animated artist, then streaming it to an SVG file would require creating both the main SVG document as well as some sort of JS helper program that can update/animate it. This sounds quite complex though. 

It might be easier to convert the artists/figure.patches/what-have-you to another JS plotting library such as [chart.js](https://www.chartjs.org/) or any of the other [seamlessly infinite number of JS plotting libs](https://css-tricks.com/svg-charting-libraries/)...

**Comment by jklymak** [comment]
Saving previous artist state can grow in memory w/o bound.  I'm not sure its possible for us to track that...

**Comment by jungerm2** [comment]
I agree. Although, any in-memory animation has that problem, which is precisely why `plt.rcParams['animation.embed_limit']` is a thing.

**Comment by jklymak** [comment]
Sure, but saving a buffer of rasters is a lot simpler than the library saving artist state for every artist.

**Comment by jungerm2** [comment]
Fair. Well, in that case, we could have the SVG backend keep that state? It would likely be less memory although it would need to figure out how things change...

**Comment by jungerm2** [comment]
I just updated my repo with another potential solution:
https://github.com/jungerm2/Matplotlib-SVG-Animation-Writer

The above-mentioned DiffWriter was a postprocessing step and as such was extremely slow. This new SVGFuncAnimation aims to be a drop-in replacement for `FuncAnimation`, instead of being a Writer. It only redraws the necessary artists and doesn't save whole frames (only the new SVG nodes) and as such is both fast and memory efficient. 

Here's a quick comparison for a plot with 1000 random points:

|   | SVGFuncAnimation + HTMLWriter | SVGFuncAnimation |
|---|---|---|
| Time Taken | 32.5 s | 1 s |  
| File Size | 31 MB | 12 MB| 

I have yet to test this more thoroughly but it's looking very promising. Take a look and let me know what you think. Thanks!

**Comment by anntzer** [comment]
From a very quick look, I think that using the artist hash() (as done in SVGFuncAnimation) may fail if an artist is drawn, *gets gc'ed*, and another artist gets allocated at the same place in memory (as that's what determines an object's `id`) -- I have vague memories of running into this kind of hash collisions a while ago.
Perhaps we should just maintain an internal counter (`itertools.count()`) and whenever an artist is *instantiated*, assign it a unique id that would serve for such bookkeeping purposes? (then the gid would default to e.g. `hash((rcParams["svg.hashsalt"], unique_id))` at draw time if gid is not set.)

**Comment by jungerm2** [comment]
That's a keen observation, that challenges one of the main assumptions in my code.

While garbage collection seems unlikely, the SVG backend maintains some internal state that assigns unique ids to _drawn_ artists. It would be great to reuse [this logic](https://github.com/matplotlib/matplotlib/blob/214def542c7ad84bcdd1d288c2fc77a4b3641a74/lib/matplotlib/backends/backend_svg.py#L629), the problem is that this id is either the artist's gid or automatically assigned. In the case where it is automatically assigned, the artist doesn't remember it, it simply gets added to the SVG doc, and the next time we draw the same artist, a new id is assigned to it. 

This is part of the reason I opted to systematically assign gids to all artists in the figure. This ensures that the id in the resulting SVG doc is consistent. But then, instead of having some meaningful ids in the SVG doc like `line2d_x`, `tick_x`, etc we have a bunch of hashes (although these "meaningful" ids aren't super helpful because the whole document is base64 encoded usually). 

In the scenario you describe (using `itertools.count()`), how would we know whenever an artist is instantiated? These get instantiated by the user's code (in setup or in the `init_func`) and thus we have no control over it, best we can do is hope these stay alive (and not gc'ed) no?

**Comment by jungerm2** [comment]
Another key assumption I made is that every artist will "reserve" some space for itself in the SVG document even if it is not drawn. This is true for lines as doing `plt.plot([], [])` will in fact create a group in the SVG document (something like `<g id=line2d_17>...</g>`), but some artists do not do this. Notably, the `Text` artist will immediately return and not even create a group if its text is empty. This means that SVGFuncAnimation won't have a handle on the text object in it's base document, which, by design, will make it throw an exception like `ValueError: Artist Text(0.02, 0.95, 'time = 0'), with gid=94936138683, not recognized.`

Redrawing the whole figure is expensive and probably should be avoided as much as possible, but it might be necessary if the base document doesn't include a group for each artist. Maybe instead of throwing this error, we could recompute the base document by saving the whole figure, but then we'd have to merge this with the previous base in order to form a base document that accommodates each artist. This sounds quite challenging in the general case, maybe there are some simplifications we could do. Or we could require every artist to draw itself the first time around (even if empty).

**Comment by anntzer** [comment]
> whenever an artist is instantiated

You can add that logic in `Artist.__init__`?

**Comment by jungerm2** [comment]
I'd rather it be self-contained and edit matplotlib's core as a last resort. Instead of the hash, I'm not using a UUID, which seems to work fine, and there's (basically) no probability of a collision anymore. 

I added a method that gathers all the SVG `<defs>` that are defined from frame to frame and adds these to the base document such that they are always available. This relies on some XML parsing which is not super pleasant, but at least there's no 3rd party import. Now animated text and scatter plots work.

I've added a few extra examples and speed/memory benchmarks to the repo. [Here's an NBViewer permalink](https://nbviewer.jupyter.org/github/jungerm2/Matplotlib-SVG-Animation-Writer/blob/3bb60a1a44b6f575968915307ab2f92bc89da820/svgfuncanim.ipynb) on which you should see the working animation and can view the page source easily.

**Comment by jungerm2** [comment]
Given the same test as in the initial post in this thread, the SVGFuncAnimation method performs rather well:
![svgfuncanim_benchmark](https://user-images.githubusercontent.com/17375657/112741398-2752f780-8f4b-11eb-9d67-d52121ac9ff1.png)


For my specific usecase (mentioned previously) which uses 4 subplots and latex math, I'm down from ~200mb to only 12mb and 30s to only 2s! I added the code to reproduce the above results on my repo.

**Comment by jungerm2** [comment]
I forked the repo and added a few non-breaking (all tests pass) changes to the SVG backend that effectively keeps track of what groups were written and delays writing the <defs> until the rendering is finalized. This enables the SVGFuncAnimation class to not depend on having to parse the doc and as such, it's faster and more reliable (the XML tree parser was messing up namespaces). I also added some proper unit tests that compare the output frame-by-frame between SVGFuncAnimation and FuncAnimation, which all pass.

What now? How should I proceed? I have yet to move the SVGFuncAnimation code into the fork for a few reasons:
1) It's kind of in-between the FuncAnimation and HTMLWriter classes and contains the necessary functionality from both, but it can't really act like them since it doesn't inherit from both FileMovieWriter and TimedAnimation (I'm not sure this would make sense). Is there a better interface I should adhere to?
2) Because this will be a new feature (hopefully), is there any standard I should adhere to beyond just [these](https://matplotlib.org/stable/devel/contributing.html#contributing-code)?

If not, I guess I'll add some docstrings and open a PR!
